File: solution2.py
# Time Complexity : push -> O(1) for all operations
# Space Complexity : push -> O(n) where n is number of elements in stack
# Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this : No


# Your code here along with comments explaining your approach

class MinStack:

    def __init__(self):
        self.stack=[]
        #maintain a min value always and add as tuple to stack
        self.minval=float('inf')
        

    def push(self, val: int) -> None:
        if not self.stack:
            self.stack.append((val,val))
        else:
            self.minval=min(val,self.stack[-1][1])
            self.stack.append((val,self.minval))
        

    def pop(self) -> None:
        self.stack.pop()
        
        

    def top(self) -> int:
        return self.stack[-1][0]
        

    def getMin(self) -> int:
        return self.stack[-1][1]

File: Solution2.java
// Time Complexity :O(1)
// Space Complexity :O(n)
// Did this code successfully run on Leetcode :Yes
// Any problem you faced while coding this :No

// Your code here along with comments explaining your approach
class MinStack {
    Stack<Integer> st;
    int min;

    public MinStack() {
        st= new Stack<>();
        min=Integer.MAX_VALUE;
    }
    
    public void push(int val) {
        if(val>min){
            st.push(val);
        }

        // to track previous min value if min value is popped 
        else{
            st.push(min);
            st.push(val);
            min=val;
        }
    }
    
    public void pop() {
        int temp=st.pop();
        if(temp==min){
            min=st.pop();
        }
    }
    
    public int top() {
        return st.peek();
    }
    
    public int getMin() {
        return min;
    }
}

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack obj = new MinStack();
 * obj.push(val);
 * obj.pop();
 * int param_3 = obj.top();
 * int param_4 = obj.getMin();
 */